File: Cpp/ListNode.cpp
//
//  ListNode.cpp
//  Cpp
//
//  Created by warden on 2021/8/6.
//

#include "ListNode.h"

ListNode* creatNodeList(std::vector<int> nodes) {
    ListNode *head = new ListNode(0);
    ListNode *p = head;
    for (int node:nodes) {
        p->next = new ListNode(node);
        p = p->next;
    }
    p = head->next;
    delete head;
    return p;
}

ListNode* creatCycleNodeList(std::vector<int> nodes, int pos) {
    ListNode *head = new ListNode(0);
    ListNode *p = head;
    ListNode* cycleNode = nullptr;
    for (int node:nodes) {
        p->next = new ListNode(node);
        if (pos-- == 0) cycleNode = p->next;
        p = p->next;
    }
    p->next = cycleNode;
    p = head->next;
    
    delete head;
    return p;
}

std::vector<int> getNodeList(ListNode *node) {
    std::vector<int> res;
    while (node) {
        res.push_back(node->val);
        node = node->next;
    }
    return res;
}

std::vector<std::vector<int>> getNodeList(std::vector<ListNode*> nodes) {
    std::vector<std::vector<int>> res;
    for (ListNode *node : nodes) {
        res.push_back(getNodeList(node));
    }
    return res;
}

void printNodeList(ListNode* node) {
    while (node != nullptr) {
        printf("%d", node->val);
        node = node->next;
        if (node != nullptr) {
            printf("->");
        }
    }
    printf("\n");
}

File: Swift/ListNode.swift
//
//  ListNode.swift
//  LeetCode-Swift
//
//  Created by warden on 2019/11/25.
//  Copyright © 2019 warden. All rights reserved.
//

import Foundation

class ListNode: Equatable {
    
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
    
    static func == (lhs: ListNode, rhs: ListNode) -> Bool {
        var n1: ListNode? = lhs
        var n2: ListNode? = rhs
        while n1 != nil && n2 != nil {
            if n1?.val != n2?.val {
                return false;
            }
            n1 = n1?.next
            n2 = n2?.next
        }
        return n1 == nil && n2 == nil;
    }
    
    var nodes: [Int] {
        var nodes_: [Int] = []
        var node: ListNode? = self
        while let n = node {
            nodes_.append(n.val)
            node = node?.next
        }
        return nodes_
    }
}

// 生成节点链表
func generateListNode(_ nodes: [Int]) -> ListNode? {
    let head = ListNode(0)
    var p = head
    for node in nodes {
        p.next = ListNode(node)
        p = p.next!
    }
    return head.next
}

// 生成节点链表
func generateListNode(_ nodes: [[Int]]) -> [ListNode?] {
    return nodes.map(generateListNode(_:))
}

// 打印节点链表
func printListNode(_ node: ListNode?) {
    if node == nil {
        return
    }
    var printNode = node;
    var arr = [String]()
    while printNode != nil {
        arr.append(String(printNode!.val))
        printNode = printNode?.next
    }
    print(arr)
}
